### PR TITLE
User authentication

### PR DESCRIPTION
**Pull request by TEMESGEN-G25**
## Summary

- Install CanCanCan in your project.
- Add a `role` column to the users table. Remember to use a migration for this.
- A user can delete a post if it is theirs or if they have an admin role (column `role` has value `"admin"`). Use CanCanCan for this authorization.
    - For that you need to implement the post deleting functionality. Add the "Delete" button to the view and make sure that only authorized users can see it.
- A user can delete a comment if it is theirs or if they have an admin role (column `role` has value `"admin"`). Use CanCanCan for this authorization.
    - For that you need to implement the comment deleting functionality. Add the "Delete" button to the view and make sure that only authorized users can see it.

### CODE REVIEW

File: app/models/ability.rb [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+    # class of the resource.
+    #
+    # The third argument is an optional hash of conditions to further filter the
+    # objects.
+    # For example, here the user can only update published articles.
+    #
+    #   can :update, Article, published: true
+    #
+    # See the wiki for details:
+    # https://github.com/CanCanCommunity/cancancan/blob/develop/docs/define_check_abilities.md
```
Comment (danmainah) [inline app/models/ability.rb]: - [x] Kindly remove the commented code, it's not good practice to have commented code in your codebase.

File: app/models/ability.rb [hunk 2]
```diff
@@ -0,0 +1,38 @@
+class Ability
+  include CanCan::Ability
+
+  def initialize(user)
+    user ||= User.new
+
+    if user.is? :admin
+      can :manage, :all, Post
+    else
```
Comment (danmainah) [inline app/models/ability.rb:10]: - [ ] Kindly add a rule that allows authors to delete their posts or comments.

File: app/controllers/comments_controller.rb [hunk 3]
```diff
@@ -18,11 +18,11 @@ def set_comment
...
-    @comment.user = current_user
+    @comment.user_id = current_login.id
     @comment.post = @post
     if @comment.save
       flash[:success] = 'Comment created successfully'
-      redirect_to user_post_path(current_user, @post)
+      redirect_to user_post_path(current_login, @post)
     else
       flash.now[:error] = 'Error: Comment could not be created'
     end
```
Comment (danmainah) [inline app/controllers/comments_controller.rb:28]: - [x] Kindly add the destroy action for your commentsController.Its a project requirement.

**Review by danmainah — CHANGES_REQUESTED** [review]
Hi @NatiG25 and @abiodunraheem, 

Good job so far!
There are some issues that you still need to work on to go to the next project but you are almost there!

## Highlights

✔️ No linter errors
✔️ Github flow was correctly used


# Required Changes ♻️
_Check the comments under the review._

## Optional suggestions
_Every comment with the [OPTIONAL] prefix is not crucial enough to stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better._


Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s).

File: app/models/ability.rb (hunk 1 above)
Comment (sja-thedude) [reply app/models/ability.rb]: Implemented 👍

File: app/models/ability.rb (hunk 2 above)
Comment (sja-thedude) [reply app/models/ability.rb:10]: Implemented 👍

File: app/controllers/comments_controller.rb (hunk 3 above)
Comment (sja-thedude) [reply app/controllers/comments_controller.rb:28]: Implemented 👍

**Review by sja-thedude — APPROVED** [review]
## Project Approved 👍 

Hi @NatiG25, This is @sja-thedude your code reviewer.

Your project is complete!  There is nothing else to say other than... it's time to merge it :shipit: 
Congratulations! 🎉 

# Highlights:
- All linters passing
- All requirements completed
- Professional readme

giphy

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me [@sja-thedude] in your question so I can receive the notification.

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561)  you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

**Comment by TEMESGEN-G25** [comment]
@danmainah @sja-thedude Thank you for all of your reviews.

**Pull request merged**